classes/output/login.php

```diff
@@ -49,51 +53,84 @@ public static function output_login_options($wantsurl) {
+        // We need to check this after calculating the IDP data to ensure its current.
+        $cache = cache::make('local_login', 'renderedlogin');
+        $cachehash = $cache->get('idphash');
+        $idphash = sha1(json_encode($idplist));
+        $cacheval = $cache->get('html');
+        if ($cachehash === $idphash && !empty($cacheval)) {
+            return $cacheval;
+        }
+
         if (!empty($config->showmanual)) {
+            $manual = new \stdClass();
+
             // Now display link to manual login page.
             $urlparams = [
                 'noredirect' => 1,
                 'passive' => 'off' // Prevent Saml2 from triggering passive check on manual login page.
             ];
-            $manualloginurl = new \moodle_url('/login/index.php', $urlparams);
+            $manual->manualurl = new \moodle_url('/login/index.php', $urlparams);
 
-            $output .= \html_writer::start_tag('div', array('class' => 'idp-login container-fluid manual'));
             if (!empty($config->beforemanualtext)) {
-                $output .= \html_writer::span(format_text($config->beforemanualtext), 'beforemanual');
+                $manual->beforemanual = format_text($config->beforemanualtext);
             }
-            $name = !empty($config->custommanualtext) ? format_string($config->custommanualtext)
-            : get_string('manuallogin', 'local_login');
-            $attributes = ['class' => 'btn btn-secondary btn-block', 'title' => $name];
-            $output .= \html_writer::link($manualloginurl, $name, $attributes);
-            $output .= \html_writer::end_tag('div');
+
+            $name = !empty($config->custommanualtext) ? format_string($config->custommanualtext) : get_string('manuallogin', 'local_login');
+            $manual->manualname = $name;
+
+            $context->manual = $manual;
         }
 
-        $output .= \html_writer::end_tag('div');
+        $context->header = format_text($config->headertext);
+        $context->footer = format_text($config->footertext);
 
-        return $output;
+        // Try to load the template from settings.
+        $template = $config->template ?? '';
+        if (empty($template)) {
+            $template = file_get_contents($CFG->dirroot . '/local/login/templates/login-minimal.mustache.default');
+        }
+
+        // Custom mustache engine to load a string into context.
+        $engine = new Mustache_Engine([
+            'escape' => 's',
+            'pragmas' => [\Mustache_Engine::PRAGMA_BLOCKS],
+        ]);
+
+        try {
+            $html = $engine->render($template, $context);
+        } catch (Mustache_Exception $e) {
+            // If something went wrong, fall back to the base template so there is *something*.
+            // If it explodes here, there are bigger problems.
+            $html = $engine->render(file_get_contents($CFG->dirroot . '/local/login/templates/login-minimal.mustache.default'), $context);
```

Comment:
Since you reference `$CFG->dirroot . '/local/login/templates/login-minimal.mustache.default'` a couple of times - maybe break this out to a const or a one liner function e.g. `get_minimal_template`